tket/src/Architecture/Architecture.cpp

```diff
@@ -24,37 +25,62 @@
 
 namespace tket {
 
-// basic implementation that works off same prior assumptions
-// TODO: Update this for more mature systems of multi-qubit gates
 bool Architecture::valid_operation(
     const Op_ptr& op, const std::vector<Node>& uids) const {
-  if (op->get_desc().is_box() ||
-      (op->get_type() == OpType::Conditional &&
-       static_cast<const Conditional&>(*op).get_op()->get_desc().is_box()))
+  if (is_box_type(op->get_type())) {
     return false;
 
-  if (uids.size() ==
-      1) {  // TODO: for simple case here this should probably not pass if
-            // node_exists[uids[0]] == FALSE, but should be fine for now?
-    return true;
-  } else if (uids.size() == 2) {
-    if (this->node_exists(uids[0]) && this->node_exists(uids[1]) &&
-        (this->edge_exists(uids[0], uids[1]) ||
-         this->edge_exists(uids[1], uids[0]))) {
-      return true;
+  } else if (op->get_type() == OpType::Conditional) {
+    if (static_cast<const Conditional&>(*op).get_op()->get_desc().is_box()) {
+      return false;
     }
-  } else if (uids.size() == 3) {
-    bool con_0_exists =
-        (this->edge_exists(uids[0], uids[1]) ||
-         this->edge_exists(uids[1], uids[0]));
-    bool con_1_exists =
-        (this->edge_exists(uids[2], uids[1]) ||
-         this->edge_exists(uids[1], uids[2]));
-    if (this->node_exists(uids[0]) && this->node_exists(uids[1]) &&
-        this->node_exists(uids[2]) && con_0_exists && con_1_exists) {
-      return true;
+    if (uids.size() == 1) {
+      return this->node_exists(uids[0]);
+    }
+    if (uids.size() == 2) {
+      if (this->node_exists(uids[0]) && this->node_exists(uids[1]) &&
+          (this->edge_exists(uids[0], uids[1]) ||
+           this->edge_exists(uids[1], uids[0]))) {
+        return true;
+      }
+    }
+  } else if (is_single_qubit_type(op->get_type())) {
+    if (uids.size() == 1) {
+      if (this->node_exists(uids[0])) {
```

Comment:
@sjdilkes Should this be included in this PR or should I just return true in all single qubits cases independent of the placement of the qubit? If this is included a few c++ tests ( and python test, but I have not checked them in details yet)  need to be changed, see below.